diff --git a/Core/Libraries/Source/debug/debug_stack.cpp b/Core/Libraries/Source/debug/debug_stack.cpp
index e40e573e..438be1bf 100644
--- a/Core/Libraries/Source/debug/debug_stack.cpp
+++ b/Core/Libraries/Source/debug/debug_stack.cpp
@@ -27,6 +27,7 @@
 // Stack walker
 //////////////////////////////////////////////////////////////////////////////
 #include "_pch.h"
+#include "stringex.h"
 #include <imagehlp.h>
 
 // Definitions to allow run-time linking to the dbghelp.dll functions.
@@ -250,8 +251,7 @@ void DebugStackwalk::Signature::GetSymbol(unsigned addr,
 
     char *p=strrchr(symbolBuffer,'\\'); // use filename only, strip off path
     p=p?p+1:symbolBuffer;
-    strncpy(bufMod,p,sizeMod);
-    bufMod[sizeMod-1]=0;
+    strlcpy(bufMod,p,sizeMod);
   }
   if (relMod)
     *relMod=addr-modBase;
@@ -266,8 +266,7 @@ void DebugStackwalk::Signature::GetSymbol(unsigned addr,
     DWORD displacement;
     if (gDbg._SymGetSymFromAddr((HANDLE)GetCurrentProcessId(),addr,&displacement,symPtr))
     {
-      strncpy(bufSym,symPtr->Name,sizeSym);
-      bufSym[sizeSym-1]=0;
+      strlcpy(bufSym,symPtr->Name,sizeSym);
       if (relSym)
         *relSym=displacement;
     }
@@ -288,8 +287,7 @@ void DebugStackwalk::Signature::GetSymbol(unsigned addr,
     {
       char *p=strrchr(line.FileName,'\\'); // use filename only, strip off path
       p=p?p+1:line.FileName;
-      strncpy(bufFile,p,sizeFile);
-      bufFile[sizeFile-1]=0;
+      strlcpy(bufFile,p,sizeFile);
       if (linePtr)
         *linePtr=line.LineNumber;
       if (relLine)
